specs/backlog/access-control.rst

```diff
@@ -0,0 +1,189 @@
+..
+ This work is licensed under a Creative Commons Attribution 3.0 Unported
+ License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+
+======================================
+Access Control for Ironic Node Actions
+======================================
+
+Include the URL of your StoryBoard story (which should have an `rfe`` tag):
+
+https://storyboard.openstack.org/#!/story/XXXXXXX
+
+Right now Ironic is not multi-tenant; anyone with API access to one node
+has access to all nodes. However there is demand for providing finer-grained
+access control to nodes. This spec describes a solution that uses a custom
+policy engine to do so.
+
+
+
+Problem description
+===================
+
+There are a variety of use cases where this capability is needed:
+
+* An Ironic deployment consists of nodes from multiple hardware Owners.
+  Each Owner needs to be able to manage their own nodes through the Ironic
+  API, while being unable to manage others' nodes.
+
+* An Ironic deployment wishes to grant specific Users the capability of
+  performing specific Ironic API actions on a node (setting the node's
+  power state) while preventing others (setting node properties).
+
+
+
+Proposed change
+===============
+
+The Ironic API already has a way of controlling access to the REST API:
+https://github.com/openstack/ironic/blob/master/ironic/common/policy.py defines
+a policy engine that is used throughout the REST API. We propose creating an
+extended version of this policy engine for use solely by the Node API. This
+policy engine still relies upon the Oslo policy file, but also enables the
+following access:
+
+* Allows an Owner of a node to automatically pass any policy check if they
+  own the node.
+
+* Allows a User of a node to pass a policy check if they have the appropriate
+  entry in a policy/node/project mapping table.
+
+The latter may be expanded to specify a policy set - another table that maps
+to a list of policies - rather than a single policy. This enables easy reuse
+of a group of policies.
+
+This alternative policy engine is enough to control access for most API
+functions - except for list functions. For those, we propose updating the
+node list query to filter the result for non-admins.
+
+A switch in the configuration would control whether you're using the "classic"
+Ironic policy engine versus this new fine-grained policy engine.
```

Comment:
Do you think this is necessary? Assuming we're agreed on the "no change to default behavior", then couldn't we use the new policy engine in all cases?